File: WeekendWarriors/Questionnaires/Questionnaires/SemanticAnalysis/SemanticAnalyzer.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Questionnaires.SemanticAnalysis
{
    public class SemanticAnalyzer
    {
        private QLContext Context = new QLContext();

        public SemanticAnalyzer()
        {

        }

        public Run.Result AnalyzeForm(AST.Form form)
        {
            return AnalyzeAstNode(form);
        }

        protected Run.Result AnalyzeAstNode(AST.INode node)
        {
            // Get and check question declarations
            var Results = new Run.DeclarationValidator().Analyze(node, Context);
            if (!Results.IsError()) // Only apply type checking if the declaration validator passed
            {
                var TypeCheckerResult = new Run.TypeChecker().Analyze(node, Context);
                Results.Combine(TypeCheckerResult);
            }
            
            return Results;
        }
    }
}

File: Hatt/taxform/src/main/java/org/uva/hatt/taxform/typechecker/SemanticAnalyzer.java
package org.uva.hatt.taxform.typechecker;

import org.uva.hatt.taxform.ast.nodes.Form;
import org.uva.hatt.taxform.typechecker.messages.Messages;

public class SemanticAnalyzer {

    private final Messages messages;
    private final TypeChecker typeChecker;
    private final CircularDependencyChecker circularDependencyChecker;

    public SemanticAnalyzer() {
        this.messages = new Messages();
        this.typeChecker = new TypeChecker(messages);
        this.circularDependencyChecker = new CircularDependencyChecker(messages);
    }

    public Messages analyze(Form form) {
        typeChecker.visit(form);
        circularDependencyChecker.visit(form);

        return messages;
    }
}
